### ISSUE TITLE
[Bug]: Error on VimEnter when using custom highlights

### ISSUE THREAD

**Issue by booperlv** [akinsho/bufferline.nvim#220]
### What happened?

When opening neovim, I am greeted by an error message:
```
Error detected while processing VimEnter Autocommands for "*":
E5108: Error executing lua vim/shared.lua:219: after the second argument
: expected table, got string
stack traceback:
        vim/shared.lua:219: in function 'tbl_extend'
        vim/shared.lua:223: in function 'tbl_extend'
        vim/shared.lua:223: in function 'tbl_deep_extend'
        ...k/packer/start/bufferline.nvim/lua/bufferline/config.lua:90:
in function 'merge'
        ...k/packer/start/bufferline.nvim/lua/bufferline/config.lua:553:
 in function 'apply'
        ...ite/pack/packer/start/bufferline.nvim/lua/bufferline.lua:1202
: in function '__load'
        [string ":lua"]:1: in main chunk
```
Following the lines, it looked like it was an issue with custom highlights.
And so,, disabling/commenting my custom highlights out of my config worked.

### What did you expect to happen?

For it to launch properly, without any errors.

### Additional Information

I have not found any specific value in highlights that causes this, I haven't tried commenting out every single one yet as that seems to be unlikely to be the cause -  but what I know is just from a quick scroll in the codebase so ¯\\\_(ツ)_/¯

Here's my config for bufferline though (pardon for the mess, haven't cleaned the file up in a while)
```lua
require "bufferline".setup {
  options = {
    offsets = {{filetype = "NvimTree", text = ""}},
    buffer_close_icon = "",
    modified_icon = "",
    close_icon = " ",
    left_trunc_marker = " ",
    right_trunc_marker = " ",
    show_buffer_icons = true,
    show_buffer_close_icons = false,
    view = "multiwindow",
    separator_style = {"",""},
    indicator_icon = "",
  },

  highlights = {
    fill = {
      guifg = {attribute="fg",highlight="TabLine"},
      guibg = {attribute="bg",highlight="TabLine"}
    },
    background = {
      guifg = {attribute="fg",highlight="TabLine"},
      guibg = {attribute="bg",highlight="TabLine"}
    },

    buffer_selected = {
      guifg = {attribute="fg",highlight="TabLineSel"},
      guibg = {attribute="bg",highlight="TabLineSel"},
      gui = "none"
    },
    buffer_visible = {
      guifg = {attribute="fg",highlight="TabLine"},
      guibg = {attribute="bg",highlight="TabLine"}
    },
  
    tab_selected = {
      guifg = {attribute="fg",highlight="TabLineSel"},
      guibg = {attribute="bg",highlight="TabLineSel"}
    },
    tab = {
      guifg = {attribute="fg",highlight="TabLine"},
      guibg = {attribute="bg",highlight="TabLine"}
    },
  
    duplicate_selected = {
      guifg = {attribute="fg",highlight="TabLineSel"},
      guibg = {attribute="bg",highlight="TabLineSel"},
      gui = "italic",
    },
    duplicate_visible = {
      guifg = {attribute="fg",highlight="TabLine"},
      guibg = {attribute="bg",highlight="TabLine"},
      gui = "italic",
    },
    duplicate = {
      guifg = {attribute="fg",highlight="TabLine"},
      guibg = {attribute="bg",highlight="TabLine"},
      gui = "italic",
    },

    modified = {
      guifg = {attribute="fg",highlight="TabLine"},
      guibg = {attribute="bg",highlight="TabLine"}
    },
    modified_selected = {
      guifg = {attribute="fg",highlight="TabLineSel"},
      guibg = {attribute="bg",highlight="TabLineSel"}
    },
    modified_visible = {
      guifg = {attribute="fg",highlight="TabLine"},
      guibg = {attribute="bg",highlight="TabLine"}
    },

    separator = {
      guifg = {attribute="bg",highlight="TabLine"},
      guibg = {attribute="bg",highlight="TabLine"}
    },
    separator_selected = {
      guifg = {attribute="bg",highlight="TabLineSel"},
      guibg = {attribute="bg",highlight="TabLineSel"}
    },
    separator_visible = {
      guifg = {attribute="bg",highlight="TabLine"},
      guibg = {attribute="bg",highlight="TabLine"}
    },
    indicator_selected = {
      guifg = {attribute="bg",highlight="TabLine"},
      guibg = {attribute="bg",highlight="TabLine"}
    },

  }
}

```

Fortunately was able to pinpoint which commit had caused it through wget-ing recent commits though :P

### commit

65a24314a2ace829356a3b7e5c791ef3e19c9ed9

**Comment by booperlv** [comment]
Tried to debug this myself, but couldn't really solve the issue. Most I found is that removing line 94 had stopped the error, but reset the config to default when changing colorschemes. Oddly enough, it says that self.__original is a string in the error, but when I tried to `print(vim.inspect(self.__original), type(self.__original))` the output was `} table`.

Tried following the events happening in the error message, and in the apply function, I placed `print(vim.inspect(defaults), type(defaults))` the output was `} table` again.

This is unfortunately outside of my ability, hope someone with more experience can get this fixed!

**Comment by akinsho** [comment]
@booperlv thanks for raising this, I'll take a look as soon as I can. I'm not really sure how is possible that the original config could be a string.

**Comment by akinsho** [comment]
@booperlv I've just tried copying your entire config and replicating this issue and it doesn't reproduce for me. Can you try to isolate this issue using a minimal init.lua i.e. only load this plugin and your colorscheme with the config above and see if it reproduces? since I don't think the setup you provided on it's own will reproduce the issue

**Comment by booperlv** [comment]
Oh man, was wondering what was wrong since nothing looked to be off. I'll do some investigating in my config and hopefully find either a solution or a minimal config to reproduce the issue. Thanks for looking into it! I'll update here with whatever I find

**Comment by booperlv** [comment]
Did some debugging and this init.lua could reproduce the issue.

```lua
vim.cmd("packadd packer.nvim")
local packer = require("packer")

local use = packer.use

return packer.startup(
	function()
		use {
      'wbthomason/packer.nvim',
    }
		use {
	    'akinsho/bufferline.nvim',
      requires = {'kyazdani42/nvim-web-devicons'},
	    config = function()
        require "bufferline".setup {
          options = {
            offsets = {{filetype = "NvimTree", text = ""}},
            buffer_close_icon = "",
            modified_icon = "",
            close_icon = " ",
            left_trunc_marker = " ",
            right_trunc_marker = " ",
            show_buffer_icons = true,
            show_buffer_close_icons = false,
            view = "multiwindow",
            separator_style = {"",""},
            indicator_icon = "",
          },

          highlights = {
            fill = {
              guifg = {attribute="fg",highlight="TabLine"},
              guibg = {attribute="bg",highlight="TabLine"}
            },
            background = {
              guifg = {attribute="fg",highlight="TabLine"},
              guibg = {attribute="bg",highlight="TabLine"}
            },

            buffer_selected = {
              guifg = {attribute="fg",highlight="TabLineSel"},
              guibg = {attribute="bg",highlight="TabLineSel"},
              gui = "none"
            },
            buffer_visible = {
              guifg = {attribute="fg",highlight="TabLine"},
              guibg = {attribute="bg",highlight="TabLine"}
            },

            tab_selected = {
              guifg = {attribute="fg",highlight="TabLineSel"},
              guibg = {attribute="bg",highlight="TabLineSel"}
            },
            tab = {
              guifg = {attribute="fg",highlight="TabLine"},
              guibg = {attribute="bg",highlight="TabLine"}
            },

            duplicate_selected = {
              guifg = {attribute="fg",highlight="TabLineSel"},
              guibg = {attribute="bg",highlight="TabLineSel"},
              gui = "italic",
            },
            duplicate_visible = {
              guifg = {attribute="fg",highlight="TabLine"},
              guibg = {attribute="bg",highlight="TabLine"},
              gui = "italic",
            },
            duplicate = {
              guifg = {attribute="fg",highlight="TabLine"},
              guibg = {attribute="bg",highlight="TabLine"},
              gui = "italic",
            },

            modified = {
              guifg = {attribute="fg",highlight="TabLine"},
              guibg = {attribute="bg",highlight="TabLine"}
            },
            modified_selected = {
              guifg = {attribute="fg",highlight="TabLineSel"},
              guibg = {attribute="bg",highlight="TabLineSel"}
            },
            modified_visible = {
              guifg = {attribute="fg",highlight="TabLine"},
              guibg = {attribute="bg",highlight="TabLine"}
            },

            separator = {
              guifg = {attribute="bg",highlight="TabLine"},
              guibg = {attribute="bg",highlight="TabLine"}
            },
            separator_selected = {
              guifg = {attribute="bg",highlight="TabLineSel"},
              guibg = {attribute="bg",highlight="TabLineSel"}
            },
            separator_visible = {
              guifg = {attribute="bg",highlight="TabLine"},
              guibg = {attribute="bg",highlight="TabLine"}
            },
            indicator_selected = {
              guifg = {attribute="bg",highlight="TabLine"},
              guibg = {attribute="bg",highlight="TabLine"}
            },
          }
        }
      end
    }
  end
)
```

I honestly don't know if this is a bigger issue with my computer or what - I don't have another machine to test on. If this does not reproduce the issue on your side: It might be an issue with something in my .local/share/nvim or the likes, in which case I might close this issue since it would be my fault then - and then add my solution here if I ever find one. In any case, sorry for the bother!

**Comment by akinsho** [comment]
@booperlv just tried it out with a few tweaks i.e. not using packer, shouldn't make a difference, One thing I noticed though is  that your minimal example doesn't use `termguicolors` which is a necessary setting for using this plugin. Tbh I don't know if it relates to this issue since I tried with and without it and in either case it doesn't reproduce the bug.

What version of neovim are you one? I'm testing on the latest nightly? I'm wondering if maybe the behaviour of the `tbl_deep_extend` function might be different in your version also is there a chance that you might have overwritten the function or something?

**Comment by booperlv** [comment]
@akinsho the version I'm using is stable, with the whole info being
```
NVIM v0.5.0
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/bin/cc -D_FORTIFY_SOURCE=2 -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -O2 -DNDEBUG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/build/neovim/src/neovim-0.5.0/build/config -I/build/neovim/src/neovim-0.5.0/src -I/usr/include -I/build/neovim/src/neovim-0.5.0/build/src/nvim/auto -I/build/neovim/src/neovim-0.5.0/build/include
Compiled by builduser

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```
I just downloaded this from the Arch package manager.

On a side note,   I got to a super minimal config that reproduces the issue on my machine:
```
vim.cmd("packadd packer.nvim")
local packer = require("packer")

local use = packer.use

return packer.startup(
function()
  use {
    'wbthomason/packer.nvim',
  }
  use {
    'akinsho/bufferline.nvim',
    requires = {'kyazdani42/nvim-web-devicons'},
    config = function()
      require "bufferline".setup {
        highlights = {
          fill = {
            guifg = {attribute="fg",highlight="TabLine"},
            guibg = {attribute="bg",highlight="TabLine"}
          }
        }
      }
    end
  }
end
)
```
Apparently, removing the values inside changing fill to `fill = {}` would stop the issue.

This is a pretty weird issue. Wonder what's wrong on my side?

**Comment by booperlv** [comment]
Fortunately have a nightly-appimage script I made from a while ago, and that fixed the issue! I downloaded the latest appimage `NVIM v0.6.0-dev+254-g413e86869` and after a +PackerSync it just worked from the get-go. Guess the behavior did change in between these versions? Will the plugin use something else for now?

**Comment by akinsho** [comment]
Okay @booperlv I've managed to reproduce this issue now, it definitely relates to a change of behaviour using vim.tbl_deep_extend in nvim-0.5 vs nvim-0.6+ 😓 really annoying since I have to build some sort of shim or duplicate the logic from nvim-0.6 in this plugin till 0.6 becomes stable

**Comment by akinsho** [comment]
This should be fixed by #222 gonna optimistically close this out. let me know if it isn't fixed though 🤞🏿

**Comment by iovis** [comment]
Seems fixed on my end 🎉
